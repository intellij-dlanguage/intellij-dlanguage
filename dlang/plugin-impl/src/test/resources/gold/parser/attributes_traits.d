@('c') string s;
pragma(msg, __traits(getAttributes, s)); // prints tuple('c')

template Tuple (T...)
{
    alias Tuple = T;
}

enum EEE = 7;
@("hello") struct SSS { }
@(3) { @(4) @EEE @SSS int foo; }

alias TP = Tuple!(__traits(getAttributes, foo));

pragma(msg, TP); // prints tuple(3, 4, 7, (SSS))
pragma(msg, TP[2]); // prints 7


TP[3] a; // a is declared as an SSS

pragma(msg, __traits(getAttributes, typeof(a))); // prints tuple("hello")