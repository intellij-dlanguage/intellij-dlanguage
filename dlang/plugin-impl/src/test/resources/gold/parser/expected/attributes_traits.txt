D Language File
  DLanguageDeclarationImpl(DECLARATION)
    DLanguageAttributeImpl(ATTRIBUTE)
      DLanguageAtAttributeImpl(AT_ATTRIBUTE)
        PsiElement(DlangTokenType.@)('@')
        PsiElement(DlangTokenType.()('(')
        DLanguageArgumentListImpl(ARGUMENT_LIST)
          DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
            DLanguageCmpExpressionImpl(CMP_EXPRESSION)
              DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                  PsiElement(DlangTokenType.CHARACTER_LITERAL)(''c'')
        PsiElement(DlangTokenType.))(')')
    DLanguageVariableDeclarationImpl(VARIABLE_DECLARATION)
      DLanguageTypeImpl(TYPE)
        DLanguageType_2Impl(TYPE_2)
          DLanguageTypeIdentifierPartImpl(TYPE_IDENTIFIER_PART)
            DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
              IDENTIFIER
                PsiElement(DlangTokenType.ID)('string')
      DECLARATOR
        IDENTIFIER
          PsiElement(DlangTokenType.ID)('s')
      PsiElement(DlangTokenType.;)(';')
  DLanguageDeclarationImpl(DECLARATION)
    DLanguagePragmaDeclarationImpl(PRAGMA_DECLARATION)
      DLanguagePragmaDeclarationImpl(PRAGMA_DECLARATION)
        DLanguagePragmaExpressionImpl(PRAGMA_EXPRESSION)
          PsiElement(DlangTokenType.pragma)('pragma')
          PsiElement(DlangTokenType.()('(')
          IDENTIFIER
            PsiElement(DlangTokenType.ID)('msg')
          PsiElement(DlangTokenType.,)(',')
          DLanguageArgumentListImpl(ARGUMENT_LIST)
            DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
              DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                  DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                    DLanguageTraitsExpressionImpl(TRAITS_EXPRESSION)
                      PsiElement(DlangTokenType.__traits)('__traits')
                      PsiElement(DlangTokenType.()('(')
                      IDENTIFIER
                        PsiElement(DlangTokenType.ID)('getAttributes')
                      PsiElement(DlangTokenType.,)(',')
                      DLanguageTemplateArgumentListImpl(TEMPLATE_ARGUMENT_LIST)
                        DLanguageTemplateArgumentImpl(TEMPLATE_ARGUMENT)
                          DLanguageTypeImpl(TYPE)
                            DLanguageType_2Impl(TYPE_2)
                              DLanguageTypeIdentifierPartImpl(TYPE_IDENTIFIER_PART)
                                DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                  IDENTIFIER
                                    PsiElement(DlangTokenType.ID)('s')
                      PsiElement(DlangTokenType.))(')')
          PsiElement(DlangTokenType.))(')')
        PsiElement(DlangTokenType.;)(';')
    PsiComment(DlangTokenType.LINE_COMMENT)('// prints tuple('c')')
  DLanguageDeclarationImpl(DECLARATION)
    TEMPLATE_DECLARATION
      PsiElement(DlangTokenType.template)('template')
      IDENTIFIER
        PsiElement(DlangTokenType.ID)('Tuple')
      DLanguageTemplateParametersImpl(TEMPLATE_PARAMETERS)
        PsiElement(DlangTokenType.()('(')
        DLanguageTemplateParameterListImpl(TEMPLATE_PARAMETER_LIST)
          TEMPLATE_PARAMETER
            DLanguageTemplateTupleParameterImpl(TEMPLATE_TUPLE_PARAMETER)
              IDENTIFIER
                PsiElement(DlangTokenType.ID)('T')
              PsiElement(DlangTokenType....)('...')
        PsiElement(DlangTokenType.))(')')
      PsiElement(DlangTokenType.{)('{')
      DLanguageDeclarationImpl(DECLARATION)
        DLanguageAliasDeclarationImpl(ALIAS_DECLARATION)
          PsiElement(DlangTokenType.alias)('alias')
          ALIAS_INITIALIZER
            IDENTIFIER
              PsiElement(DlangTokenType.ID)('Tuple')
            PsiElement(DlangTokenType.=)('=')
            DLanguageTypeImpl(TYPE)
              DLanguageType_2Impl(TYPE_2)
                DLanguageTypeIdentifierPartImpl(TYPE_IDENTIFIER_PART)
                  DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                    IDENTIFIER
                      PsiElement(DlangTokenType.ID)('T')
          PsiElement(DlangTokenType.;)(';')
      PsiElement(DlangTokenType.})('}')
  DLanguageDeclarationImpl(DECLARATION)
    DLanguageVariableDeclarationImpl(VARIABLE_DECLARATION)
      DLanguageAutoDeclarationImpl(AUTO_DECLARATION)
        DLanguageStorageClassImpl(STORAGE_CLASS)
          PsiElement(DlangTokenType.enum)('enum')
        AUTO_DECLARATION_PART
          IDENTIFIER
            PsiElement(DlangTokenType.ID)('EEE')
          PsiElement(DlangTokenType.=)('=')
          DLanguageInitializerImpl(INITIALIZER)
            DLanguageNonVoidInitializerImpl(NON_VOID_INITIALIZER)
              DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                  DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                    DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                      PsiElement(DlangTokenType.INTEGER_LITERAL)('7')
        PsiElement(DlangTokenType.;)(';')
  DLanguageDeclarationImpl(DECLARATION)
    DLanguageAttributeImpl(ATTRIBUTE)
      DLanguageAtAttributeImpl(AT_ATTRIBUTE)
        PsiElement(DlangTokenType.@)('@')
        PsiElement(DlangTokenType.()('(')
        DLanguageArgumentListImpl(ARGUMENT_LIST)
          DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
            DLanguageCmpExpressionImpl(CMP_EXPRESSION)
              DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                  PsiElement(DlangTokenType.DOUBLE_QUOTED_STRING)('"hello"')
        PsiElement(DlangTokenType.))(')')
    STRUCT_DECLARATION
      PsiElement(DlangTokenType.struct)('struct')
      IDENTIFIER
        PsiElement(DlangTokenType.ID)('SSS')
      DLanguageStructBodyImpl(STRUCT_BODY)
        PsiElement(DlangTokenType.{)('{')
        PsiElement(DlangTokenType.})('}')
  DLanguageDeclarationImpl(DECLARATION)
    DLanguageAttributeImpl(ATTRIBUTE)
      DLanguageAtAttributeImpl(AT_ATTRIBUTE)
        PsiElement(DlangTokenType.@)('@')
        PsiElement(DlangTokenType.()('(')
        DLanguageArgumentListImpl(ARGUMENT_LIST)
          DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
            DLanguageCmpExpressionImpl(CMP_EXPRESSION)
              DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                  PsiElement(DlangTokenType.INTEGER_LITERAL)('3')
        PsiElement(DlangTokenType.))(')')
    PsiElement(DlangTokenType.{)('{')
    DLanguageDeclarationImpl(DECLARATION)
      DLanguageAttributeImpl(ATTRIBUTE)
        DLanguageAtAttributeImpl(AT_ATTRIBUTE)
          PsiElement(DlangTokenType.@)('@')
          PsiElement(DlangTokenType.()('(')
          DLanguageArgumentListImpl(ARGUMENT_LIST)
            DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
              DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                  DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                    PsiElement(DlangTokenType.INTEGER_LITERAL)('4')
          PsiElement(DlangTokenType.))(')')
      DLanguageAttributeImpl(ATTRIBUTE)
        DLanguageAtAttributeImpl(AT_ATTRIBUTE)
          PsiElement(DlangTokenType.@)('@')
          IDENTIFIER
            PsiElement(DlangTokenType.ID)('EEE')
      DLanguageAttributeImpl(ATTRIBUTE)
        DLanguageAtAttributeImpl(AT_ATTRIBUTE)
          PsiElement(DlangTokenType.@)('@')
          IDENTIFIER
            PsiElement(DlangTokenType.ID)('SSS')
      DLanguageVariableDeclarationImpl(VARIABLE_DECLARATION)
        DLanguageTypeImpl(TYPE)
          DLanguageType_2Impl(TYPE_2)
            DLanguageBuiltinTypeImpl(BUILTIN_TYPE)
              PsiElement(DlangTokenType.int)('int')
        DECLARATOR
          IDENTIFIER
            PsiElement(DlangTokenType.ID)('foo')
        PsiElement(DlangTokenType.;)(';')
    PsiElement(DlangTokenType.})('}')
  DLanguageDeclarationImpl(DECLARATION)
    DLanguageAliasDeclarationImpl(ALIAS_DECLARATION)
      PsiElement(DlangTokenType.alias)('alias')
      ALIAS_INITIALIZER
        IDENTIFIER
          PsiElement(DlangTokenType.ID)('TP')
        PsiElement(DlangTokenType.=)('=')
        DLanguageTypeImpl(TYPE)
          DLanguageType_2Impl(TYPE_2)
            DLanguageTypeIdentifierPartImpl(TYPE_IDENTIFIER_PART)
              DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                DLanguageTemplateInstanceImpl(TEMPLATE_INSTANCE)
                  IDENTIFIER
                    PsiElement(DlangTokenType.ID)('Tuple')
                  DLanguageTemplateArgumentsImpl(TEMPLATE_ARGUMENTS)
                    PsiElement(DlangTokenType.!)('!')
                    PsiElement(DlangTokenType.()('(')
                    DLanguageTemplateArgumentListImpl(TEMPLATE_ARGUMENT_LIST)
                      DLanguageTemplateArgumentImpl(TEMPLATE_ARGUMENT)
                        DLanguageTypeImpl(TYPE)
                          DLanguageType_2Impl(TYPE_2)
                            DLanguageTraitsExpressionImpl(TRAITS_EXPRESSION)
                              PsiElement(DlangTokenType.__traits)('__traits')
                              PsiElement(DlangTokenType.()('(')
                              IDENTIFIER
                                PsiElement(DlangTokenType.ID)('getAttributes')
                              PsiElement(DlangTokenType.,)(',')
                              DLanguageTemplateArgumentListImpl(TEMPLATE_ARGUMENT_LIST)
                                DLanguageTemplateArgumentImpl(TEMPLATE_ARGUMENT)
                                  DLanguageTypeImpl(TYPE)
                                    DLanguageType_2Impl(TYPE_2)
                                      DLanguageTypeIdentifierPartImpl(TYPE_IDENTIFIER_PART)
                                        DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                          IDENTIFIER
                                            PsiElement(DlangTokenType.ID)('foo')
                              PsiElement(DlangTokenType.))(')')
                    PsiElement(DlangTokenType.))(')')
      PsiElement(DlangTokenType.;)(';')
  DLanguageDeclarationImpl(DECLARATION)
    DLanguagePragmaDeclarationImpl(PRAGMA_DECLARATION)
      DLanguagePragmaDeclarationImpl(PRAGMA_DECLARATION)
        DLanguagePragmaExpressionImpl(PRAGMA_EXPRESSION)
          PsiElement(DlangTokenType.pragma)('pragma')
          PsiElement(DlangTokenType.()('(')
          IDENTIFIER
            PsiElement(DlangTokenType.ID)('msg')
          PsiElement(DlangTokenType.,)(',')
          DLanguageArgumentListImpl(ARGUMENT_LIST)
            DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
              DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                  DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                    DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                      IDENTIFIER
                        PsiElement(DlangTokenType.ID)('TP')
          PsiElement(DlangTokenType.))(')')
        PsiElement(DlangTokenType.;)(';')
    PsiComment(DlangTokenType.LINE_COMMENT)('// prints tuple(3, 4, 7, (SSS))')
  DLanguageDeclarationImpl(DECLARATION)
    DLanguagePragmaDeclarationImpl(PRAGMA_DECLARATION)
      DLanguagePragmaDeclarationImpl(PRAGMA_DECLARATION)
        DLanguagePragmaExpressionImpl(PRAGMA_EXPRESSION)
          PsiElement(DlangTokenType.pragma)('pragma')
          PsiElement(DlangTokenType.()('(')
          IDENTIFIER
            PsiElement(DlangTokenType.ID)('msg')
          PsiElement(DlangTokenType.,)(',')
          DLanguageArgumentListImpl(ARGUMENT_LIST)
            DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
              DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                  DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                    DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                      IDENTIFIER
                        PsiElement(DlangTokenType.ID)('TP')
                  DLanguageIndexExpressionImpl(INDEX_EXPRESSION)
                    PsiElement(DlangTokenType.[)('[')
                    DLanguageIndexImpl(INDEX)
                      DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                        DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                          DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                            DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                              PsiElement(DlangTokenType.INTEGER_LITERAL)('2')
                    PsiElement(DlangTokenType.])(']')
          PsiElement(DlangTokenType.))(')')
        PsiElement(DlangTokenType.;)(';')
    PsiComment(DlangTokenType.LINE_COMMENT)('// prints 7')
  DLanguageDeclarationImpl(DECLARATION)
    DLanguageVariableDeclarationImpl(VARIABLE_DECLARATION)
      DLanguageTypeImpl(TYPE)
        DLanguageType_2Impl(TYPE_2)
          DLanguageTypeIdentifierPartImpl(TYPE_IDENTIFIER_PART)
            DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
              IDENTIFIER
                PsiElement(DlangTokenType.ID)('TP')
            PsiElement(DlangTokenType.[)('[')
            DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
              DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                  DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                    PsiElement(DlangTokenType.INTEGER_LITERAL)('3')
            PsiElement(DlangTokenType.])(']')
      DECLARATOR
        IDENTIFIER
          PsiElement(DlangTokenType.ID)('a')
      PsiElement(DlangTokenType.;)(';')
    PsiComment(DlangTokenType.LINE_COMMENT)('// a is declared as an SSS')
  DLanguageDeclarationImpl(DECLARATION)
    DLanguagePragmaDeclarationImpl(PRAGMA_DECLARATION)
      DLanguagePragmaDeclarationImpl(PRAGMA_DECLARATION)
        DLanguagePragmaExpressionImpl(PRAGMA_EXPRESSION)
          PsiElement(DlangTokenType.pragma)('pragma')
          PsiElement(DlangTokenType.()('(')
          IDENTIFIER
            PsiElement(DlangTokenType.ID)('msg')
          PsiElement(DlangTokenType.,)(',')
          DLanguageArgumentListImpl(ARGUMENT_LIST)
            DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
              DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                  DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                    DLanguageTraitsExpressionImpl(TRAITS_EXPRESSION)
                      PsiElement(DlangTokenType.__traits)('__traits')
                      PsiElement(DlangTokenType.()('(')
                      IDENTIFIER
                        PsiElement(DlangTokenType.ID)('getAttributes')
                      PsiElement(DlangTokenType.,)(',')
                      DLanguageTemplateArgumentListImpl(TEMPLATE_ARGUMENT_LIST)
                        DLanguageTemplateArgumentImpl(TEMPLATE_ARGUMENT)
                          DLanguageTypeImpl(TYPE)
                            DLanguageType_2Impl(TYPE_2)
                              DLanguageTypeofExpressionImpl(TYPEOF_EXPRESSION)
                                PsiElement(DlangTokenType.typeof)('typeof')
                                PsiElement(DlangTokenType.()('(')
                                DLanguageExpressionImpl(EXPRESSION)
                                  DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                    DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                      DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                        DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                          DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                            IDENTIFIER
                                              PsiElement(DlangTokenType.ID)('a')
                                PsiElement(DlangTokenType.))(')')
                      PsiElement(DlangTokenType.))(')')
          PsiElement(DlangTokenType.))(')')
        PsiElement(DlangTokenType.;)(';')
    PsiComment(DlangTokenType.LINE_COMMENT)('// prints tuple("hello")')